lib/wallaroo/core/sink/connector_sink/connector_sink_checkpoint_rollback.pony

```diff
@@ -0,0 +1,465 @@
+  =>
+    _invalid_call(__loc.method_name()); Fail()
+
+  fun ref enter(sink: ConnectorSink ref) =>
+    None
+
+  fun ref phase1_abort(sink: ConnectorSink ref, txn_id: String) =>
+    _invalid_call(__loc.method_name()); Fail()
+
+  fun ref phase1_commit(sink: ConnectorSink ref, txn_id: String) =>
+    _invalid_call(__loc.method_name()); Fail()
+
+  fun ref prepare_for_rollback(sink: ConnectorSink ref) =>
+    _CpRbTransition(this, _CpRbPreparedForRollback, sink)
+
+  fun ref rollback(sink: ConnectorSink ref,
+    barrier_token: CheckpointBarrierToken) =>
+    _invalid_call(__loc.method_name()); Fail()
+
+  fun ref rollbackresume_barrier_complete(sink: ConnectorSink ref) =>
+    _invalid_call(__loc.method_name()); Fail()
+
+  fun _invalid_call(method_name: String) =>
+    @l(Log.crit(), Log.conn_sink(),
+      "Invalid call to %s on _CpRbOps state %s".cstring(),
+      method_name.cstring(), name().cstring())
+
+primitive _CpRbTransition
+  fun apply(curr: _CpRbOps, next: _CpRbOps, sink: ConnectorSink ref) =>
+    let debug: LogSevCat = Log.make_sev_cat(Log.debug(), Log.conn_sink())
+
+    @ll(debug, "CpRbTransition:: %s -> %s".cstring(),
+      curr.name().cstring(), next.name().cstring())
+    // We must update sink's _cprb pointer before calling .enter()!
+    // Otherwise, if .enter() also updates the _cprb pointer, then
+    // a pointer update will get clobbered & lost.
+    sink._update_cprb_member(next)
+    next.enter(sink)
+    @ll(debug, "CpRbTransition:: %s -> %s DONE".cstring(),
+      curr.name().cstring(), next.name().cstring())
+
+class _CpRbAbortCheckpoint is _CpRbOps
+  let _debug: LogSevCat = Log.make_sev_cat(Log.debug(), Log.conn_sink())
+  var _checkpoint_to_abort: (None | CheckpointBarrierToken)
+  var _abort_sent: Bool = false
+  let _send_phase1: Bool
+
+  fun name(): String => __loc.type_name()
+
+  new create(checkpoint_to_abort: (None | CheckpointBarrierToken)) =>
+    _checkpoint_to_abort = checkpoint_to_abort
+    _send_phase1 = match checkpoint_to_abort
+      | None =>
+        true
+      else
+        false
+      end
+
+  fun ref enter(sink: ConnectorSink ref) =>
+    sink.cprb_send_advertise_status(false)
+    match sink.cprb_check_sink_phase_partial_barrier()
+    | None =>
+      None
+    | let cbt: CheckpointBarrierToken =>
+      @l(Log.info(), Log.conn_sink(),
+        "_CpRbAbortCheckpoint: detected shear with token %s".cstring(),
+          cbt.string().cstring())
+      match _checkpoint_to_abort
+      | let cpta: CheckpointBarrierToken =>
+        if cpta != cbt then
+          @l(Log.crit(), Log.conn_sink(),
+            "Expected %s but partial barrier is %s".cstring(),
+            _checkpoint_to_abort.string().cstring(), cbt.string().cstring())
+          Fail()
+        end
+      end
+    end
+    _ChangeSinkPhaseQueueMsgsForwardTokens(sink where shear_risk = true)
+    _is_checkpoint_id_known(sink)
+
+  fun ref _is_checkpoint_id_known(sink: ConnectorSink ref) =>
```

Comment:
This seems like it should return a Bool. Can you rename to reflect what's happening?

Reply:
Sure